[JS/Web] Add uniforms to Einsum

satyajandhyala opened the pull request:
### Description
Add uinforms to Einsum



### Motivation and Context
<!-- - Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here. -->
Improve performance.

fs-eire commented on js/web/lib/wasm/jsep/webgpu/ops/einsum.ts:
```diff
@@ -249,21 +254,34 @@ const createEinsumProgramInfo = (inputs: readonly TensorView[], einsumEquation:
+      inputDependencies: enableInputShapesUniforms.map((enableShapeUniform) => enableShapeUniform ? 'rank' : 'dims')
+    },
     getRunData: () => ({
       outputs: [{dims: outputShape, dataType: inputs[0].dataType}],
```
This is incorrect.

When input dependencies are "rank" instead of "dims", the return value of function `getRunData()` should only depends on the function's parameter for inputs. For example:

this is wrong:
```ts
 getRunData: () => ({
      outputs: [{dims: outputShape, dataType: inputs[0].dataType}], // outputShape and inputs[0] come from closure and are for this time.
      ...
```

This is correct:
```ts
 getRunData: (inputs) => ({
      outputs: [{dims: calcOutputShape(inputs, attribute), dataType: inputs[0].dataType}], // always come from parameters of function "getRunData"
      ...
```

fs-eire replied:
The conclusion above is not true - the current code can ensure the correctness.

The ProgramInfo is designed as - 
the following 2 parts should be light-weighted
- the ProgramInfo object should be generated in low cost
- the getRunData() function should be lightweighted

the 'getShaderSource' function may contain more calculations as necessary. it will only be called once unless cache miss.

satyajandhyala replied:
Moved code from createEinsumProgramInfo to getShaderSource function.

qjia7 reviewed (commented):
L237 https://github.com/microsoft/onnxruntime/pull/18531/files#diff-174cff6a2add89e890c3b0b225c89a3f45057105364adbfb0a578318157e845dR237 
For `${symbol} < ${einsumEquation.symbolToInfo.get(symbol)?.dimValue}`, the `dimValue` should be a uniform instead of constant. Otherwise, how to distinguish below two cases?
case 1:
equation: "ij,jk->ik", inputs[0].dims = [2, 3], inputs[0].dims = [3, 4], outputShape = [2, 4]
case 2:
equation: "ij,jk->ik", inputs[0].dims = [2, 6], inputs[0].dims = [6, 4], outputShape = [2, 4].

Another small concern is that using `einsumEquation.equation` as shaderCache's hint is a little strict. For example, below two cases can share same shader source. But due to the symbol is different, they will be compiled twice:
case 1:
equation: "ij,jk->ik", inputs[0].dims = [2, 3], inputs[0].dims = [3, 4], outputShape = [2, 4]
case 2:
equation: "ik,kj->ij", inputs[0].dims = [2, 3], inputs[0].dims = [3, 4], outputShape = [2, 4]
You don't need to resolve this one in this PR since it may need many changes to refactor. Just point it out. You can think about it in future. Other parts look good to me. Thanks.

satyajandhyala commented:
> L237 https://github.com/microsoft/onnxruntime/pull/18531/files#diff-174cff6a2add89e890c3b0b225c89a3f45057105364adbfb0a578318157e845dR237 For `${symbol} < ${einsumEquation.symbolToInfo.get(symbol)?.dimValue}`, the `dimValue` should be a uniform instead of constant. Otherwise, how to distinguish below two cases? case 1: equation: "ij,jk->ik", inputs[0].dims = [2, 3], inputs[0].dims = [3, 4], outputShape = [2, 4] case 2: equation: "ij,jk->ik", inputs[0].dims = [2, 6], inputs[0].dims = [6, 4], outputShape = [2, 4].
> 
> Another small concern is that using `einsumEquation.equation` as shaderCache's hint is a little strict. For example, below two cases can share same shader source. But due to the symbol is different, they will be compiled twice: case 1: equation: "ij,jk->ik", inputs[0].dims = [2, 3], inputs[0].dims = [3, 4], outputShape = [2, 4] case 2: equation: "ik,kj->ij", inputs[0].dims = [2, 3], inputs[0].dims = [3, 4], outputShape = [2, 4] You don't need to resolve this one in this PR since it may need many changes to refactor. Just point it out. You can think about it in future. Other parts look good to me. Thanks.

Good findings!
Fixed the first problem resulting in incorrect computation and added a test case.

qjia7 reviewed (approved):
LGTM, thanks.
